File: ch09/App_Code/VB/Listing9-13.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.ModelBinding

Namespace VB
    
    Public Class AggregateValueProvider
        Implements IValueProvider
        Implements IUnvalidatedValueProvider
        Private ReadOnly _valueProviders As New List(Of IUnvalidatedValueProvider)()

        Public Sub New(modelBindingExecutionContext As ModelBindingExecutionContext)
            _valueProviders.Add(New FormValueProvider(modelBindingExecutionContext))
            _valueProviders.Add(New QueryStringValueProvider(modelBindingExecutionContext))
        End Sub

        Public Function ContainsPrefix(prefix As String) As Boolean Implements IValueProvider.ContainsPrefix
            Return _valueProviders.Any(Function(vp) vp.ContainsPrefix(prefix))
        End Function

        Public Function GetValue(key As String) As ValueProviderResult Implements IValueProvider.GetValue
            Return GetValue(key, False)
        End Function

        Public Function GetValue(key As String, skipValidation As Boolean) As ValueProviderResult Implements IUnvalidatedValueProvider.GetValue
            Return _valueProviders.[Select](Function(vp) vp.GetValue(key, skipValidation)).LastOrDefault()
        End Function
    End Class

    Public Class AggregateValueAttribute
        Inherits ValueProviderSourceAttribute
        Public Overrides Function GetValueProvider(modelBindingExecutionContext As ModelBindingExecutionContext) As IValueProvider
            Return New AggregateValueProvider(modelBindingExecutionContext)
        End Function
    End Class
End Namespace

File: ch09/App_Code/CS/Listing9-13.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.ModelBinding;


public class AggregateValueProvider : IValueProvider, IUnvalidatedValueProvider
{
    private readonly List<IUnvalidatedValueProvider> _valueProviders = new List<IUnvalidatedValueProvider>();

    public AggregateValueProvider(ModelBindingExecutionContext modelBindingExecutionContext)
    {
        _valueProviders.Add(new FormValueProvider(modelBindingExecutionContext));                
        _valueProviders.Add(new QueryStringValueProvider(modelBindingExecutionContext));                
    }

    public bool ContainsPrefix(string prefix)
    {
        return _valueProviders.Any(vp => vp.ContainsPrefix(prefix));
    }

    public ValueProviderResult GetValue(string key)
    {
        return GetValue(key, false);
    }

    public ValueProviderResult GetValue(string key, bool skipValidation)
    {
        return _valueProviders.Select(vp => vp.GetValue(key, skipValidation))
            .LastOrDefault();
    }
}

public class AggregateValueAttribute : ValueProviderSourceAttribute
{
    public override IValueProvider GetValueProvider(ModelBindingExecutionContext modelBindingExecutionContext)
    {
        return new AggregateValueProvider(modelBindingExecutionContext);
    }
}
